services/skus/appstore.go

```diff
@@ -0,0 +1,308 @@
+	//     - can be used to create a new order or prepare to working in multi-device mode;
+	// - x.val.NotificationType == appstore.NotificationTypeV2Revoke && x.val.Subtype == "":
+	//     - a family member lost access to the subscription.
+
+	return x.shouldRenew() || x.shouldCancel()
+}
+
+// shouldRenew reports whether the ntf is about renewal.
+//
+// TODO: Update to a lookup table instead.
+func (x *appStoreSrvNotification) shouldRenew() bool {
+	switch {
+	// Auto-renew.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidRenew && x.val.Subtype == "":
+		return true
+
+	// Billing has recovered.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidRenew && x.val.Subtype == appstore.SubTypeV2BillingRecovery:
+		return true
+
+	// Resubscribed after cancellation.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidChangeRenewalStatus && x.val.Subtype == appstore.SubTypeV2AutoRenewEnabled:
+		return true
+
+	// Resubscribed after expiration or a family member gained access after resubscription.
+	case x.val.NotificationType == appstore.NotificationTypeV2Subscribed && x.val.Subtype == appstore.SubTypeV2Resubscribe:
+		return true
+
+	default:
+		return false
+	}
+}
+
+// shouldCancel reports whether the ntf is about cancellation.
+//
+// TODO: Update to a lookup table instead.
+func (x *appStoreSrvNotification) shouldCancel() bool {
+	switch {
+	// Cancellation or refund.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidChangeRenewalStatus && x.val.Subtype == appstore.SubTypeV2AutoRenewDisabled:
+		return true
+
+	// Extra events relating to cancellations or refunds.
+	//
+	// Need to make sure the order is cancelled.
+	// Apple processed a refund.
+	case x.val.NotificationType == appstore.NotificationTypeV2Refund && x.val.Subtype == "":
+		return true
+
+	// Expiration after user's cancellation.
+	case x.val.NotificationType == appstore.NotificationTypeV2Expired && x.val.Subtype == appstore.SubTypeV2Voluntary:
+		return true
+
+	// Expiration after billing retry ended without recovery.
+	case x.val.NotificationType == appstore.NotificationTypeV2Expired && x.val.Subtype == appstore.SubTypeV2BillingRetry:
+		return true
+
+	// Cancellation after price increase.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidChangeRenewalStatus && x.val.Subtype == "":
+		return true
+
+	default:
+		return false
+	}
+}
+
+func (x *appStoreSrvNotification) effect() string {
+	switch {
+	case x.shouldRenew():
+		return "renew"
+	case x.shouldCancel():
+		return "cancel"
+	default:
+		return "skip"
+	}
+}
+
+func parseAppStoreSrvNotification(vrf *assnCertVerifier, spayload string) (*appStoreSrvNotification, error) {
+	certs, err := extractASSNCerts(spayload)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(certs) != 3 {
+		return nil, errInvalidASSNCertNum
+	}
+
+	if err := vrf.verify(certs[2], certs[1]); err != nil {
```

Comment:
verified this is being called each time before process so that we're separating the verification logic of the token and ASSN Certs from the application logic of the renewal/TxnInfo. LGTM